runtime/runtime/src/lib.rs

```diff
@@ -1097,32 +1114,45 @@ impl Runtime {
         if current_gas_price > action_receipt.gas_price {
-            // In a rare scenario, when the current gas price is higher than the purchased gas
-            // price, the difference is subtracted from the refund. If the refund doesn't have
-            // enough balance to cover the difference, then the remaining balance is considered
-            // the deficit and it's reported in the stats for the balance checker.
-            gas_deficit_amount = safe_gas_to_balance(
-                current_gas_price - action_receipt.gas_price,
-                result.gas_burnt,
-            )?;
-            if gas_balance_refund >= gas_deficit_amount {
-                gas_balance_refund -= gas_deficit_amount;
-                gas_deficit_amount = 0;
-            } else {
-                gas_deficit_amount -= gas_balance_refund;
-                gas_balance_refund = 0;
-            }
+            // price increased, burning resulted in a deficit
+            gas_deficit = value_delta;
         } else {
-            // Refund for the difference of the purchased gas price and the current gas price.
-            gas_balance_refund = safe_add_balance(
-                gas_balance_refund,
-                safe_gas_to_balance(
-                    action_receipt.gas_price - current_gas_price,
-                    result.gas_burnt,
-                )?,
-            )?;
+            // price decreased, burning resulted in a surplus
+            gas_surplus = value_delta;
+        };
+
+        if config.fees.refund_gas_price_changes {
+            // In this version of the protocol, we rebalance the gas
+            // deficit/surplus by subtracting/adding to the paid out refund.
+            // Ideally, this should result in deficit = 0 and surplus = 0;
+            if gas_deficit > 0 {
+                // In a rare scenario, when the current gas price is higher than the purchased gas
+                // price, the difference is subtracted from the refund. If the refund doesn't have
+                // enough balance to cover the difference, then the remaining balance is considered
+                // the deficit and it's reported in the stats for the balance checker.
+                if gas_balance_refund >= gas_deficit {
+                    gas_balance_refund -= gas_deficit;
+                    gas_deficit = 0;
+                } else {
+                    gas_deficit -= gas_balance_refund;
+                    gas_balance_refund = 0;
+                }
+            } else if gas_surplus > 0 {
+                // Refund for the difference of the purchased gas price and the current gas price.
+                gas_balance_refund = safe_add_balance(gas_balance_refund, gas_surplus)?;
+                gas_surplus = 0;
+            }
+            debug_assert_eq!(
+                gas_surplus, 0,
+                "must not run a surplus with `refund_gas_price_changes` enabled"
+            );
```

Comment:
This part is iffy. Why are we prescribing logic in release builds to a condition which panics in debug builds? Why not have it be a no-op and optionally panic in debug builds?

Reply:
I'd say this whole `generate_refund_receipts` function should be left as is, a new function added to implement the new behaviour and the caller should switch between the two functions based on the configuration variable/protocol version/etc. This is now much more palatable given that we have an option to remove the old function once the new code has been running in production for some time.

Reply:
That's sounds like a great idea! I will refactor accordingly.

Reply:
> This part is iffy. Why are we prescribing logic in release builds to a condition which panics in debug builds? Why not have it be a no-op and optionally panic in debug builds?

I'm not quite following. What is your suggestion to use instead of `debug_assert!`?

My intention was to explicitly state the post-condition and check it in tests and fuzzers, while keeping it out of release builds for (maybe unfounded)  performance reasons. Happy to change it to something else though if you have a better idea.